rpc_new/client.ts

```diff
@@ -0,0 +1,152 @@
+import { Deferred, deferred } from "../deps/std/async.ts";
+import { getOrInit } from "../util/mod.ts";
+import * as msg from "./messages.ts";
+import { Provider, ProviderListener } from "./provider/base.ts";
+import { ProviderHandlerError, ProviderSendError } from "./provider/errors.ts";
+import * as U from "./util.ts";
+
+export class Client<
+  DiscoveryValue = any,
+  SendErrorData = any,
+  HandlerErrorData = any,
+  CloseErrorData = any,
+> {
+  providerRef;
+  pendingCalls: Record<string, Deferred<unknown>> = {};
+  pendingSubscriptions: SubscriptionListeners<SendErrorData, HandlerErrorData> = {};
+  activeSubscriptions: SubscriptionListeners<SendErrorData, HandlerErrorData> = {};
+  activeSubscriptionByMessageId: Record<string, string> = {};
+  subscriptionStates = new Map<string, WeakMap<new() => any, any>>();
+
+  constructor(
+    readonly provider: Provider<DiscoveryValue, SendErrorData, HandlerErrorData, CloseErrorData>,
+    readonly discoveryValue: DiscoveryValue,
+  ) {
+    this.providerRef = provider(discoveryValue, this.#listener);
+  }
+
+  #listener: ProviderListener<SendErrorData, HandlerErrorData> = (e) => {
+    if (e instanceof Error) {
+      for (const id in this.pendingCalls) {
+        const pendingCall = this.pendingCalls[id]!;
+        pendingCall.resolve(e);
+        delete this.pendingCalls[id];
+        this.pendingSubscriptions[id]!(e);
+        delete this.pendingSubscriptions[id];
+      }
+      for (const id in this.activeSubscriptions) {
+        this.activeSubscriptions[id]!(e);
+        delete this.activeSubscriptions[id];
+        this.subscriptionStates.delete(id);
+      }
+    } else if (e.id) {
+      const pendingCall = this.pendingCalls[e.id]!;
+      pendingCall.resolve(e);
+      delete this.pendingCalls[e.id];
+      if (this.pendingSubscriptions[e.id]) {
+        if (e.error) {
+          this.pendingSubscriptions[e.id]!(e);
+        } else {
+          this.activeSubscriptions[e.result] = this.pendingSubscriptions[e.id]!;
+          this.activeSubscriptionByMessageId[e.id] = e.result;
+        }
+        delete this.pendingSubscriptions[e.id];
+      }
+    } else if (e.params) {
+      this.activeSubscriptions[e.params.subscription]?.(e);
```

Comment:
@harrysolovay I just reverted this to `?.` because after calling `.stop()` from inside a subscription, the inner WebSocket could still emit an event for the stopped subscription.